src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+                    if (tabSplit.size() == 4) {
+                        microarrayIdentifierLines.add(microarrayResponseLine + "\n");
+                    }
+                }
+
+                // Write data to file.
+                if (!microarrayIdentifierLines.isEmpty()) {
+                    File speciesMicroarrayFile = createNewFile(this.destination + biomartSpeciesName + "_microarray");
+                    for (String probeIdLine : microarrayIdentifierLines) {
+                        Files.write(Paths.get(speciesMicroarrayFile.toString()), probeIdLine.getBytes(), StandardOpenOption.APPEND);
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Attempts to create a connection to Biomart with the supplied query
+     * @param queryString String of Biomart query
+     * @return Set<String> Biomart response data.
+     * @throws Exception Can be caused by the query URL not being valid, or the data not existing for the species.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws Exception {
+        // Create connection to Biomart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == 200) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // Biomart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < 5) {
+                        Thread.sleep(10000);
+                        logger.warn("Biomart query failed. Trying again...");
+                        return queryBiomart(queryString, retryCount);
+                    }
+                    throw new Exception("Biomart query failed with message: " + line);
+                } else {
+                    biomartResponseLines.add(line);
+                }
+            }
+            return biomartResponseLines;
+
+        } else {
+            retryCount++;
+            if (retryCount < 5) {
+                Thread.sleep(10000);
+                logger.warn("Unable to connect to Biomart. Trying again...");
+                return queryBiomart(queryString, retryCount);
+            }
+            throw new Exception("Unable to connect to Biomart with URL: " + queryString);
```

Comment:
Since this is from not getting a 200 response after retrying, this exception might be good to use: https://docs.oracle.com/javase/8/docs/api/javax/xml/ws/http/HTTPException.html

Reply:
Actually, looking closer, this might be more appropriate: https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpException.html since the javax.xml.ws.http.HTTPException above can't have an error message set to it (it only receives the status code).

You could do something like this:

```
throw new HttpException(
    "Unable to connect to Biomart (" + 
        biomartConnection.getResponseCode() + ": " + biomartConnection.getResponseMessage() + 
    ") with URL: " + queryString
);
```